bom/application/pom.xml

```diff
@@ -4677,6 +4852,23 @@
                     </exclusion>
                 </exclusions>
             </dependency>
+            <!-- TODO current infinispan-client is using 6.0.8 and isn't breaking with the following constraints
+            <dependency>
+                <groupId>org.hibernate.search</groupId>
+                <artifactId>hibernate-search-engine</artifactId>
+                <version>${hibernate-search.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.hibernate.search</groupId>
+                <artifactId>hibernate-search-mapper-pojo-base</artifactId>
+                <version>${hibernate-search.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.hibernate.search</groupId>
+                <artifactId>hibernate-search-util-common</artifactId>
+                <version>${hibernate-search.version}</version>
+            </dependency>
```

Comment:
@yrodiere could you have a look at this when you're back? It looks like we cannot enforce Hibernate Search's version properly when using Infinispan.

Reply:
See this comment specifically: https://github.com/quarkusio/quarkus/pull/26565#issuecomment-1176584494

Reply:
@gsmet wdyt about merging it a figuring out the hibernate-search/infinispan compatibility separately? Given that the rest of the constraints aren't related to that.

Reply:
Last time I checked there wasn't a more recent Infinispan version for us to upgrade to, unfortunately.

Never breaking the SPI (2) is a great one :) And if you do, a safe way would be to coordinate releases, so that there is combination of Hibernate Search and Infinispan that works, which is (3).

Shading is an option, although it **may** complicate CVE tracking and patch delivery. But maybe not that much.

As to how wise downgrading the HS version is, it depends on how often people will run into this issue, basically how critical it is to have a working combination of the two. I am not sure for long it has been broken and whether anybody has run into it yet.

Reply:
Thanks for your answer. To clarify, when I said none of the options are good, I meant it ;)

> Never breaking the SPI (2) is a great one :) And if you do, a safe way would be to coordinate releases, so that there is combination of Hibernate Search and Infinispan that works, which is (3).

Yeah unfortunately that's not a realistic option. Infinispan is using very... let's say low-level SPIs, they are very close to the internals and not evolving those would very likely mean not evolving Hibernate Search. 6.1 would have been close to impossible to get out without those SPI changes.

I mean, even if we stop changing SPIs in minors (which is in line with our [compatibility policy](https://hibernate.org/community/compatibility-policy/#spi), by the way), at some point we'll have to switch to another major, and then it will just be the same problem all over again. So we can try to limit backwards-incompatible breaks, but that's really just a mitigation. 

Maybe mitigation would be enough, though. Maybe 6.1 was exceptional and it'll be easier to preserve backwards compatibility in the future.

> And if you do, a safe way would be to coordinate releases, so that there is combination of Hibernate Search and Infinispan that works, which is (3).

That's also been proven hard in the past, but yeah, we can try.

> Shading is an option, although it may complicate CVE tracking and patch delivery. But maybe not that much.

I'm not that worried about CVE in Hibernate Search / Lucene, they've been rare. But yeah that's one concern.

> As to how wise downgrading the HS version is, it depends on how often people will run into this issue, basically how critical it is to have a working combination of the two. I am not sure for long it has been broken and whether anybody has run into it yet.

FWIW, that's the first time I hear of this problem :)

Reply:
Despite all the complications, we shouldn't just let it go neither. Today we could document it as not being supported. Until one day we get one or more important use-cases that require both HS and Infinispan being present in the same app, at that moment we'll have to make it work somehow and starting from that moment, we will want to preserve compatibility and not have random breakages between releases. So thinking/preparing for that moment could be worth it.

@gsmet I don't think the rest of the changes should be blocked by this issue. Would you prefer to remove the comments I added about the HS/Infinispan issue from the BOM?

Reply:
> not have random breakages between releases

It's not random, it's when necessary, on minor versions, in SPIs only, and that's a clearly established policy for all Hibernate projects. Let's not assume we're breaking stuff just because we don't care: we do :)

The exact reason we're having a problem is I'd been told in the past that there would not be any Infinispan Embedded in Quarkus, and since the HotRod (non-Embedded) client  does not depend on Hibernate Search, the Hibernate Search dependency shouldn't have affected it at all.

> So thinking/preparing for that moment could be worth it.

Certainly.

I'm back from PTO, so I'll be able to look further. First step would be to determine why the Hibernate Search dependency affects Infnispan at all; I thought Infinispan Embedded had been added to Quarkus without my knowledge, but that doesn't seem to be the case? It seems the error is really just in the HotRod client, which shouldn't be depending on Hibernate Search in the first place. I'll look into that.

What should I do to reproduce the issue; just uncomment those lines, build the relevant modules and run Quarkus + Infinispan integration tests?

> Today we could document it as not being supported

Sure, documenting would be a good temporary solution.

Reply:
Yes, to reproduce uncomment the constraints and run `io.quarkus.it.infinispan.client.HealthCheckTest.testHealthCheck`, you can see the failures from the CI above. Thanks!

Reply:
Thanks.

For future reference, the problem was in the test dependencies rather than in the "main" dependencies. Guillaume addressed it here: https://github.com/quarkusio/quarkus/pull/26647